text/0000-debuginfo-based-panic.md

```diff
@@ -0,0 +1,115 @@
+- Feature Name: debuginfo_based_panic
+- Start Date: 2017-09-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Instead of generating location strings for panic information, read it at runtime from debug information. To make this work in release builds, introduce symbol files (external debug symbols) which are built in release mode by default.
+
+# Motivation
+[motivation]: #motivation
+
+By convention, recoverable errors are usually handled in Rust with the `Result` type while unrecoverable errors (usually bugs) cause a panic. When that happens, it is thus often necessary to figure out exactly what happened so the cause can be fixed.
+
+Today, `panic!()` attempts to help with that by printing a message along with the source code location where it happened:
+
+```
+thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /checkout/src/libcore/option.rs:335
+note: Run with `RUST_BACKTRACE=1` for a backtrace.
+```
+
+However, this is not very useful. A panic's root cause is rarely located within the panicking function - most commonly, it happened because a function's *preconditions were violated*.
+
+The textbook example for this is of course `unwrap()` (on both `Option` and `Result`) as well as similar interfaces like `Index::index`/`IndexMut::index_mut`. The above example refers to the location in `option.rs` where the panic happened, but the `unwrap()` implementation is just doing its job here - it was really caused by a bug that caused the user application to believe an `Option` has to be `Some` in a case where it's not. When debugging this, the given location information is utterly useless - the `unwrap()` function panicked so of course `panic!()` was called from there!
+
+When dealing with issues like this, one usually enables `RUST_BACKTRACE=1`. This generates a full stack trace with information about where the bug may have originated. When even that isn't enough, a debugger can provide valuable insights into what's really going on.
+
+The main purpose of panic location information as well as backtraces and debuggers is **debugging**: Finding and fixing the bug. Debugging code written in a compiled systems language is a hard problem but Rust is not unique here - the C ecosystem was facing exactly the same problem and debug info (in a platform-specific format like DWARF or PDB) exists to solve it. Of course, Rust is already leveraging this to support both backtraces and debuggers. But from this perspective it certainly seems very redundant to compile dedicated panic location strings when this information already exists in a binary's debug sections.
+
+Furthermore, there is no automated solution right now that could remove panic location strings from binaries. This is a problem for developers of closed-source software as it can, in some circumstances, faciliate reverse engineering.
```

Comment:
Does this RFC deal with this issue?

Reply:
The idea here is that if the standard panic handler just discards `file!()` and `line!()`, I would expect the optimizer to remove them (see https://github.com/rust-lang/rfcs/pull/2070#issuecomment-316752603). This is nice because it avoids having to deal with reworking all of those panic internals (although doing that is probably a cleaner and better solution).

Just like the design of the compiler flag, this is a detail that I wanted to gloss over until later rounds of feedback (it already took me a while to find the time to write up this first draft, no point in delaying it further for things that don't really matter when I'm just looking for feedback on the basic idea).

Reply:
How is [`std::panic::set_hook`](https://doc.rust-lang.org/std/panic/fn.set_hook.html) going to work?

Reply:
Quoting from [`PanicInfo::location()`](https://doc.rust-lang.org/std/panic/struct.PanicInfo.html#method.location):

> This method will currently always return `Some`, but this may change in future versions.

Payload still exists of course. And location can be resolved from debuginfo or just return `None` when that fails. Adding a method to `PanicInfo` that returns the entire stack trace seems like a good idea too.